### PR TITLE
Better analytics for cache [ACT-480]

### PR DESCRIPTION
**Pull request by trapacska**

### CODE REVIEW

File: models/remote_log.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+// RemoteLog ...
+type RemoteLog struct {
+	LogLevel string                 `json:"log_level"`
+	Message  string                 `json:"message"`
+	Data     map[string]interface{} `json:"data"`
+}
+
+// Event ...
+func (a RemoteLog) Event() string {
+	return "remote_log"
```
Comment (lpusok) [inline models/remote_log.go:12]: As discussed, please update Data team on the data fields be loogged in Segment, event name and user ID specifically, and to let them know this new use case.

File: service/custom_logs_post.go [hunk 2]
```diff
@@ -2,53 +2,40 @@ package service
...
-	case logLvlWarn:
-		logger.Warn(logData.Message, zap.Any("raw_json_data", logData.Data))
-	case logLvlError:
-		logger.Error(logData.Message, zap.Any("raw_json_data", logData.Data))
+	dogstatsd, err := GetClientFromContext(r.Context())
+	if err != nil {
+		return errors.WithStack(err)
 	}
 
+	dogstatsd.Track(log)
```
Comment (lpusok) [inline service/custom_logs_post.go:38]: Are the cases returning httpresponse.RespondWithBadRequestError tracked in datadog also?

File: models/remote_log.go (hunk 1 above)
Comment (Szabadember) [reply models/remote_log.go:12]: Informed András Lendvay, he approved the event name and the user id.

**Review by Szabadember — APPROVED** [review]

File: service/custom_logs_post.go (hunk 2 above)
Comment (trapacska) [reply service/custom_logs_post.go:38]: ah it is just a naming left over. it should be `segmentClient`

File: service/custom_logs_post.go [hunk 3]
```diff
@@ -2,53 +2,40 @@ package service
...
-	logger, err := zap.NewProduction()
-	if err != nil {
-		fmt.Printf("Failed to initialize logger: %s", err)
+	if len(log.LogLevel) == 0 {
+		return httpresponse.RespondWithBadRequestError(w, "Invalid request body, please provide log_level")
+	}
+	if len(log.Message) == 0 {
+		return httpresponse.RespondWithBadRequestError(w, "Invalid request body, please provide message")
+	}
+	if _, ok := log.Data["step_id"].(string); !ok {
```
Comment (godrei) [inline service/custom_logs_post.go:26]: NOTE: by having a function for getting a field from data, some duplication could be avoided:

```
func (a RemoteLog) DataField(key string) (string, bool) {
	d, ok := a.data[key].(string)
	return d, ok
}
```

also could be used here: https://github.com/bitrise-io/bitrise-step-analytics/pull/13/files#diff-ff296a0cb815ef5bd00483b0178b2d4dR21

File: service/custom_logs_post.go [hunk 4]
```diff
@@ -2,53 +2,40 @@ package service
...
+		return httpresponse.RespondWithBadRequestError(w, "Invalid request body, please provide data.step_id")
```
Comment (godrei) [inline service/custom_logs_post.go:27]: NOTE: `Invalid request body, please provide data.step_id as a string`

File: service/custom_logs_post.go (hunk 2 above)
Comment (godrei) [reply service/custom_logs_post.go:38]: @trapacska pls rename the var this case

File: service/custom_logs_post_test.go [hunk 5]
```diff
@@ -26,41 +23,49 @@ func Test_CustomLogsPostHandler(t *testing.T) {
...
 		{
-			testName:           "when request body isn't a valid JSON, it retrieves bad reqest error",
+			testName:           "error, when all required param missing",
+			requestBody:        `{}`,
 			expectedStatusCode: http.StatusBadRequest,
-			expectedBody:       `{"message":"Invalid request body, JSON decode failed"}` + "\n",
+			expectedBody:       `{"message":"Invalid request body, please provide log_level"}` + "\n",
+		},
+		{
+			testName:           "error, when log level set",
```
Comment (godrei) [inline service/custom_logs_post_test.go:42]: NOTE: the issue is not the presence of the `log level`, but the missing `message`,

might `error, missing message` would be more straightforward.

File: service/custom_logs_post_test.go [hunk 6]
```diff
@@ -78,3 +83,17 @@ func Test_CustomLogsPostHandler(t *testing.T) {
...
+
+func Test_CustomLogsPostHandlerInvalidContext(t *testing.T) {
+	handler := service.CustomLogsPostHandler
+
+	r, err := http.NewRequest("POST", "/logs", bytes.NewBuffer([]byte(`{"log_level":"warn","message":"test","data":{"step_id":"test-step-id","tag":"test-tag"}}`)))
+	require.NoError(t, err)
+
+	rr := httptest.NewRecorder()
+	internalServerError := handler(rr, r)
+	expectedInternalErr := "DogStatsD not found in Context"
```
Comment (godrei) [inline service/custom_logs_post_test.go:95]: NOTE: not the `segmentClient` is missing

**Review by godrei — CHANGES_REQUESTED** [review]
Pls check my notes

File: service/custom_logs_post_test.go (hunk 6 above)
Comment (trapacska) [reply service/custom_logs_post_test.go:95]: this is an internal log for DD, coming from GetClientFromContext

**Review by godrei — APPROVED** [review]

**Pull request merged**